scripts: ci: added twister_report_analyzer.py

gchwier opened the pull request:
Added script to parse and analyse Twister reports. Failed tests are grouped together by reason of failure, Cmake and build errors are grouped using predefined patterns.

Shared our internal script as was asked in discussion:
https://discord.com/channels/720317445772017664/733037890514321419/1336672406718255126

```
$ scripts/ci/twister_report_analyzer.py --help
usage: twister_report_analyzer.py [-h] [--error-patterns ERROR_PATTERNS] [--output OUTPUT] [--output-csv OUTPUT_CSV] [--status]
                                  [--long-summary] [-ll {DEBUG,INFO,WARNING,ERROR,CRITICAL}]
                                  inputs [inputs ...]

Analyzes Twister JSON reports

positional arguments:
  inputs                twister.json files to read

options:
  -h, --help            show this help message and exit
  --error-patterns ERROR_PATTERNS
                        text file with custom error patterns, each entry must be separated by newlines
  --output OUTPUT       output json file name, default: twister_report_summary.json
  --output-csv OUTPUT_CSV
                        output csv file name
  --status              add statuses of testsuites and testcases to the summary
  --long-summary        show all matched errors grouped by reason, otherwise show only most common errors
  -ll {DEBUG,INFO,WARNING,ERROR,CRITICAL}, --log-level {DEBUG,INFO,WARNING,ERROR,CRITICAL}

Example usage:
    To analyze errors with predefined CMake and Build error patterns, run:
    > python twister_report_analyzer.py twister_reports/*.json --long-summary
    The summary will be saved to twister_report_summary.json file unless --output option is used.
    To save error summary to CSV file, use --output-csv option (number of test files is limited to 100):
    > python twister_report_analyzer.py twister_reports/*.json --output-csv twister_report_summary.csv
    One can use --error-patterns option to provide custom error patterns file:
    > python twister_report_analyzer.py **/twister.json --error-patterns error_patterns.txt
```

Example output:
```
./twister_report_analyzer.py tmp/Twister_*
...
INFO    :  Data saved to twister_report_summary.json

Most common errors summary:
 141    ld.bfd: region overflowed
 109    undefined symbol NET_SOCKETS_POSIX_NAMES
 102    error: 'BT_LE_ADV_OPT_CONNECTABLE' is deprecated [-Werror=deprecated-declarations]
  88    error: 'MSG_DONTWAIT' undeclared here (not in a function); did you mean 'NRF_MSG_DONTWAIT'?
  25    devicetree error
...
```

gchwier committed:
scripts: ci: added twister_report_analyzer.py

Added script to parse and analyse Twister reports.
Failed tests are grouped together by reason of failure,
Cmake and build errors are grouped using predefined patterns.

Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

PerMac commented:
some comments raised during testing WG meeting:
- An extracted error might be added to as an extra entry in twister's report for a test
- Not that useful for majority of PRs, but can help with some bringing greater changes
- Can help analyzing weekly runs

nashif commented:
worth going all the way and run the script here https://github.com/zephyrproject-rtos/zephyr/blob/7254a615307d732cf5b4535baf5c6360261c03fa/.github/workflows/twister.yaml#L211 and generate the file as a artifact.

gchwier committed:
ci: twister: add twister report analysis

Added steps to analyze Twister reports and upload the results
if Twister build fails

Signed-off-by: Grzegorz Chwierut <grzegorz.chwierut@nordicsemi.no>

nashif commented on .github/workflows/twister.yaml:
```diff
@@ -243,6 +243,21 @@ jobs:
+      - name: Analyze Twister Reports
+        if: failure()
+        run: |
+          ./scripts/ci/twister_report_analyzer.py artifacts/**/twister.json
```
the script does not exist in this context, you will need to checkout the zephyr code first...

kartben replied:
Not sure a full checkout is needed for just a script, I would just wget the file I think?

fabiobaltieri replied:
plus checkout with default parameters is fairly quick, it does a shallow clone by default

gchwier replied:
I've added (copied):
```
      - name: Check out source code
        if: failure()
        uses: actions/checkout@v4
        with:
          ref: ${{ github.event.pull_request.head.sha }}
          fetch-depth: 0
          persist-credentials: false
```
hope it is OK

nashif commented:
> Not sure a full checkout is needed for just a script, I would just wget the file I think?

wget from a git repo is bad, your get into versioning problems and branch issues

gchwier commented:
Updated with options:
`--output-md OUTPUT_MD  output markdown file name to store table with errors`
and
`--platforms           add errors per platform to the summary`

To test one, one can try:
```
gh run download  13095023661
scripts/ci/twister_report_analyzer.py Unit*/**/twister.json --long-summary --platforms --output-md errors.md
```

Grouping by platforms is added to json output file, and printed on the console:
```
Errors per platform:
  36    ucans32k1sic/s32k146
  19    nrf9280pdk/nrf9280/cpuapp
  14    efm32wg_stk3800/efm32wg990f256
  10    intel_adsp/ace15_mtpm/sim
  10    intel_adsp/ace20_lnl/sim
  10    intel_adsp/ace30/ptl/sim
   9    intel_adsp/ace15_mtpm
   9    intel_adsp/ace20_lnl
...
```
